text/0000-naming-conventions.md

```diff
@@ -0,0 +1,267 @@
+<?php
+namespace Templates;
+
+class Page_Template implements Template {
+  // Specific implementation for the pages.
+}
+```
+
+The `extends` and `implements` keywords must be declared on the same line as the class name. The opening brace for the class must go in the same line as the class name.
+
+#### Properties
+
+Visibility must be declared on all properties. Properties should not be prefiexed with a single underscore to indicate protected or private visibility.
+
+```php
+<?php
+namespace Templates;
+
+class Page_Template implements Template {
+  public $template_name = 'page_template';
+}
+```
+
+#### Methods
+
+Visibility must be declared on all methods. Methods should not be prefiexed with a single underscore to indicate protected or private visibility.
+
+Method names must not be declared with a space after the method name. The opening brace must go on the same line as the method declaration, and the closing brace must go on the next line following the body.
+
+```php
+<?php
+namespace Admin;
+
+class Main {
+  private $theme_slug = 'my_theme';
+
+  public function __construct() {
+    $this->theme_slug = $theme_slug;
+  }
+
+  public function get_theme_name() {
+    return $this->theme_slug;
+  }
+
+  public function foo( $arg1, &$arg2, $arg3 = [] ) {
+    // method body.
+  }
+}
+```
+
+The method arguments should follow liberal spacing rules set by the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#space-usage).
+
+#### abstract, static and final
+
+When present, `abstract` and `final` declarations must precede the visibility declaration.
+
+When present, `static` declaration must come after the visibility declaration.
+
+```php
+<?php
+namespace Namespace_Name;
+
+abstract class Class_Name {
+
+  protected static $property;
+
+  abstract protected function foo();
+
+  final public static function bar() {
+    // method body.
+  }
+}
+```
+
+#### File naming
+
+Files must be named descriptively, using lowercase letters separated by hyphens. When writing functional code use short descriptive names
+
+```php
+content-gallery.php
+```
+
+Class file names must be based on the class name, with the underscores in the class name replaced by hyphens. There must be only one class in a class file. For example class `WP_Error` should reside in a
+
+```php
+wp-error.php
+```
+
+# Drawbacks
+
+If the core will be rewritten, the time it takes to implement this rule will be significant, plus we must wait for the core to drop the support for the PHP 5.2 which should have happened a long time ago (EOL for PHP 5.2 was 6 Jan 2011 according to [this table](http://php.net/eol.php)). Even PHP 5.6 will have security support until 31 Dec 2018 (according to [this table](http://php.net/supported-versions.php)).
+
+Another thing we should consider is file structure re organization so that it aligns with the OOP principles, but such that it doesn't break userspace. Which is also a pretty big task.
```

Comment:
That is not a consideration for naming conventions, that's a consideration for directory structure/namespacing/autoloading strategy, which is a different matter.